### ISSUE TITLE
Selecting a non existing workspace returns exit 0 with pg backend

### ISSUE THREAD

**Issue by baurmatt** [hashicorp/terraform#24427]
### Terraform Version
```
$ terraform version
Terraform v0.12.24
```

(Tests show that this is broken since 0.12.8)

### Terraform Configuration Files
```hcl
terraform {
  backend "pg" {}
}
```

### Debug Output
(As the debug is only about 18 lines, I will skip the gist)

Terraform 0.12.7:
```
syseleven@sys11services01:~/test$ TF_LOG=trace ../terraform-0.12.7 workspace select default
2020/03/23 11:46:08 [INFO] Terraform version: 0.12.7
2020/03/23 11:46:08 [INFO] Go runtime version: go1.12.9
2020/03/23 11:46:08 [INFO] CLI args: []string{"/home/syseleven/terraform-0.12.7", "workspace", "select", "default"}
2020/03/23 11:46:08 [DEBUG] Attempting to open CLI config file: /home/syseleven/.terraformrc
2020/03/23 11:46:08 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/03/23 11:46:08 [INFO] CLI command args: []string{"workspace", "select", "default"}
2020/03/23 11:46:08 [TRACE] Meta.Backend: built configuration for "pg" backend with hash value 2343951456
2020/03/23 11:46:08 [TRACE] Preserving existing state lineage "7cf29f36-3778-7f99-df7a-9147c13e1139"
2020/03/23 11:46:08 [TRACE] Preserving existing state lineage "7cf29f36-3778-7f99-df7a-9147c13e1139"
2020/03/23 11:46:08 [TRACE] Meta.Backend: working directory was previously initialized for "pg" backend
2020/03/23 11:46:08 [TRACE] backendConfigNeedsMigration: failed to decode cached config; migration codepath must handle problem: unsupported attribute "skip_schema_creation"
2020/03/23 11:46:08 [TRACE] Meta.Backend: backend configuration has changed (from type "pg" to type "pg")
Backend reinitialization required. Please run "terraform init".
Reason: Backend configuration changed for "pg"

The "backend" is the interface that Terraform uses to store state,
perform operations, etc. If this message is showing up, it means that the
Terraform configuration you're using is using a custom configuration for
the Terraform backend.

Changes to backend configurations require reinitialization. This allows
Terraform to setup the new configuration, copy existing state, etc. This is
only done during "terraform init". Please run that command now then try again.

If the change reason above is incorrect, please verify your configuration
hasn't changed and try again. At this point, no changes to your existing
configuration or state have been made.


Error: Initialization required. Please see the error message above.
```

Terraform 0.12.24:
```
syseleven@sys11services01:~/test$ TF_LOG=trace ../terraform-0.12.24 workspace select default
2020/03/23 11:47:31 [INFO] Terraform version: 0.12.24
2020/03/23 11:47:31 [INFO] Go runtime version: go1.12.13
2020/03/23 11:47:31 [INFO] CLI args: []string{"/home/syseleven/terraform-0.12.24", "workspace", "select", "default"}
2020/03/23 11:47:31 [DEBUG] Attempting to open CLI config file: /home/syseleven/.terraformrc
2020/03/23 11:47:31 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/03/23 11:47:31 [INFO] CLI command args: []string{"workspace", "select", "default"}
2020/03/23 11:47:31 [TRACE] Meta.Backend: built configuration for "pg" backend with hash value 241288904
2020/03/23 11:47:31 [TRACE] Preserving existing state lineage "7cf29f36-3778-7f99-df7a-9147c13e1139"
2020/03/23 11:47:31 [TRACE] Preserving existing state lineage "7cf29f36-3778-7f99-df7a-9147c13e1139"
2020/03/23 11:47:31 [TRACE] Meta.Backend: working directory was previously initialized for "pg" backend
2020/03/23 11:47:31 [TRACE] Meta.Backend: using already-initialized, unchanged "pg" backend configuration
2020/03/23 11:47:31 [TRACE] Meta.Backend: instantiated backend of type *pg.Backend
2020/03/23 11:47:31 [DEBUG] checking for provider in "."
2020/03/23 11:47:31 [DEBUG] checking for provider in "/home/syseleven"
2020/03/23 11:47:31 [DEBUG] checking for provisioner in "."
2020/03/23 11:47:31 [DEBUG] checking for provisioner in "/home/syseleven"
2020/03/23 11:47:31 [INFO] Failed to read plugin lock file .terraform/plugins/linux_amd64/lock.json: open .terraform/plugins/linux_amd64/lock.json: no such file or directory
2020/03/23 11:47:31 [TRACE] Meta.Backend: backend *pg.Backend does not support operations, so wrapping it in a local backend
```

### Crash Output
No crash happens.

### Expected Behavior

The command `terraform workspace select $workspace` should return an exit code `1` if the workspace doesn't exist. This was true before Terraform 0.12.7 and allowed to automatically create a workspace in CI as described in https://github.com/hashicorp/terraform/issues/16191#issuecomment-574661663 

### Actual Behavior
Since Terraform 0.12.8 the command `terraform workspace select $workspace` returns an exit code `0` for non existing workspaces.

### Steps to Reproduce
1. `echo -e "terraform {\n  backend \"pg\" {}\n}" > main.tf`
1. `terraform init -backend-config="conn_str=postgres://$username:$password@$host/$db" -input=false`
1. `terraform workspace select $workspace` # Run this with 0.12.7 and 0.12.8 (or above) to see the difference

**Comment by danieldreier** [comment]
@baurmatt thank you for reporting this. I've tried to reproduce this, and have not been able to. I scripted a trivial reproduction case in https://github.com/danieldreier/terraform-issue-reproductions/tree/master/24427 just to make sure we're trying exactly the same things.

Here's what I saw with 0.12.24:

```
bash-3.2$ ./run.sh
+ terraform-v0.12.24 init

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.null: version = "~> 2.1"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
+ terraform-v0.12.24 workspace select foo

Workspace "foo" doesn't exist.

You can create this workspace with the "new" subcommand.
```

I think this is the behavior you were expecting, but not what you saw, so I would appreciate it if you could take a look at my reproduction case and check whether it behaves the same on your system, and then if possible make a PR to reproduce it.

The only thing at all different between our environments is that I'm not using the postgresql backend. If the problem keeps happening for you, can you please check whether this is specific to that state backend, or if it happens regardless of backend?

Again - thank you for reporting this and I'm curious what you find.

**Comment by erikerikson** [comment]
Hi @danieldreier - I'm not the original reporter but given the context of #16191 I suspect that the issue relates to `new` rather than `select` which does exhibit the 0 exit code on the existence of the workspace (terraform `0.12.26`).  (It's a little hair-brained but it makes sense to run `new` before `select` since something must exist before it is selected.)  That written, I have no idea if the behavior is differential across the versions mentioned in this issue.  I landed on here from #16191 because I thought it would regard `new` which, regardless of a message that seems to indicate failure and an actual non-creation of the workspace (given that it already exists) I would have expected a non-zero exit but observed an exit of 0 and non-selection of the existing namespace.  Perhaps confounding this is that I am using an `azurerm` backend and the workspace only exists in that backend.

I resolved this in my script by running `select` first and falling back to `new` if that fails but the lack of failure of `new` was unexpected and is, in my opinion, a bug.  Discovering this bug and workarounds was unnecessary friction.  I would prefer `workspace new` to be idempotent (as suggested in #16191 - selecting the workspace if it already exists) but understand that would be a change in behavior that could have ramifications for users.  Thus, I've added my +1 on #16191 in support of an idempotency flag and I am advocating here that it is a bug for `workspace new` to fail to create but also fail to select an existing workspace and return a 0 exit code.

Perhaps you would prefer a separate issue?  Certainly @baurmatt would be appropriate in asking me not to hijack if I misperceived the circumstance.

**Comment by oed-mertenss** [comment]
We have in general a failure while initializing a new state with pg as backend on a not existing schema. 

Complete new project without existing states and new postgres schema name.
```
terraform workspace show
-> default

terraform init -backend-config=<path>
Initializing the backend...

Successfully configured the backend "pg"! Terraform will automatically
use this backend unless the backend configuration changes.

Error: No existing workspaces.

Use the "terraform workspace" command to create and select a new workspace.
If the backend already contains existing workspaces, you may need to update
the backend configuration.

terraform workspace show
-> default

terraform workspace list
-> without output (exit code 0)
```

When i look into the schema the structure table/seq is there but the data is missing. Maybe it is the same problem otherwise i should open a new ticket.

**Comment by baurmatt** [comment]
@oed-mertenss This looks exactly like the bug I've triggered! :)

@danieldreier Your reproduction case seem to use the local backend. My error (and the one @oed-mertenss mentioned) happened with PG as backend.

**Comment by danieldreier** [comment]
@baurmatt I'm going to label this as a pg backend issue, because I haven't been able to reproduce it locally. It sounds like it's specific to the pg backend rather than being general to terraform workspace management.

@erikerikson if you're seeing a general issue where creating a workspace that already exists results in exit code 0, that would be a different issue. If it's specific to just postgresql, I think including a reproduction case here is probably sufficient. 

I tested that against the local backend, and my trivial case didn't reproduce what I think you described, so I'm wondering if you're also hitting a different expression of the same underlying pg backend issue:

```
$ terraform-v0.12.29 workspace new foo
Created and switched to workspace "foo"!

You're now on a new, empty workspace. Workspaces isolate their state,
so if you run "terraform plan" Terraform will not see any existing state
for this configuration.

$ terraform-v0.12.29 workspace new foo
Workspace "foo" already exists
```

(I did the same with 0.13.0 and got the same behavior)

**Comment by danieldreier** [comment]
I've labeled this as needs-maintainer because the pg backend currently lacks a dedicated maintainer. If any of the folks here are interested in contributing in that way, feel free to reach out.

**Comment by erikerikson** [comment]
Thanks Daniel.  My case is occurring with the "azurerm" backend (terraform 0.12.26 and azurerm 2.13.0) .  I'll report it elsewhere, apologies for the noise.

**Comment by remilapeyre** [comment]
Hi @baurmatt, I looked into this and this seems specific to the `default` workspace that was not created during the initialization of the backend, hence running `terraform workspace select default` would exit early with the 0 exit code since it was the workspace we were supposed to be already using even if it did not exist. Trying to select to another workspace, for example `terraform workspace select foo`, would exit with a non zero exit code as it would actually try to switch to it and notice it does not exist.

The fact that the `default` workspace was not created was a bug that has been fixed in https://github.com/hashicorp/terraform/pull/26420 so the behavior should be correct now.
